### PR TITLE
python: tidy dependency management

### PR DESCRIPTION
**Pull request by mrpollo**
This commit introduces [Pipenv](https://pipenv.readthedocs.io/en/latest/) into the workflow allowing for
deterministic releases due to dependency pinning

* plus the unintended consequence of updating README documentation

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+scipy = "*"
+cython = "*"
+pyfftw = "*"
+pylint = "*"
+requests = "*"
+
+[dev-packages]
+
+[requires]
+python_version = "3.7"
```
[the remark this answers is not in the document]
Comment (mrpollo) [reply Pipfile]: This is my bad, I was intending to use 3.6, if you are ok with 3.7 you should run ```pipenv --three```` before everything else, I'll update the docs

[the remark this answers is not in the document]
Comment (bkueng) [reply Pipfile]: Ok, I don't want to be too restrictive here. Anything with python 3.x should just work by default.

[the remark this answers is not in the document]
Comment (mrpollo) [reply Pipfile]: What is production using? we should lock to that.

**Comment by mrpollo** [comment]
@bkueng updated docs

File: Pipfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrpollo) [reply Pipfile]: Just sent a new changeset locking to 3.4.9 and better

**Comment by pietrodn** [comment]
Pipenv is great for deterministic builds, because it solves the long-time dilemma of whether to lock package versions in `requirements.txt`.

I think it should be used as the primary package management system for the project (i.e., for CI and documentation). Pipenv is also [recommended](https://packaging.python.org/tutorials/managing-dependencies/) by the official Python guide about application dependency management. So far, so good.

However, I also had mixed experiences in using Pipenv, especially on resource-constrained systems.
Here are the problems I encountered:
1) Pipenv is a fairly large project with some dependencies, which must be installed *globally*. One may not wish to install it globally on the system; or worse, one may not have the permission to do so. `pip` is generally shipped within the Python distribution; if it's not, the builtin Python `venv` module [ensures that it gets installed](https://docs.python.org/3/library/venv.html#venv.EnvBuilder) within any newly created virtual environment, without needing root privileges.
2) As outlined by @bkueng, Pipenv is [much slower](https://chriswarrick.com/blog/2018/07/17/pipenv-promises-a-lot-delivers-very-little/#the-part-where-i-try-to-measure-times) than `pip`, because package locking takes a lot of time. Personally, some time ago I ditched Pipenv for pip+venv exactly because of this.
3) Pipenv [does not play well](https://github.com/pypa/pipenv/issues/15) with virtual environments created with `venv`, which is [the standard, suggested way](https://docs.python.org/3/library/venv.html#creating-virtual-environments) of creating virtual environments in Python 3. Both venv and Pipenv are officially recommended by the Python documentation, so this is strange. But we know Python packaging is a mess, so I guess this is ok... sort of.
4) I really don't like using `pipenv run` to execute my scripts, especially in automated, non-interactive scenarios. It [slows down application startup](https://github.com/pypa/pipenv/issues/2053), and it requires additional environment variables to be set ([1](https://pipenv.readthedocs.io/en/latest/diagnose/#valueerror-unknown-locale-utf-8), [2](https://pipenv.readthedocs.io/en/latest/diagnose/#using-pipenv-run-in-supervisor-program)). When automating things (e.g. with `systemd` services or `cron`), you have to make sure that `pipenv` is on your `$PATH` and that [the working directory is set correctly](http://www.robpeck.com/2017/11/using-pipenv-with-systemd/).
5) By default, Pipenv creates the virtual environment in a [non-deterministic folder name](https://github.com/pypa/pipenv/issues/2316), so either you set another environment variable to override this behaviour, or you're stuck with `pipenv run` to run your script non-interactively.

I think that all these problems may be solved by generating a `requirements.txt` file (very easy, with Pipenv: `pipenv lock --requirements`), and by committing it.
In this way, the user can choose the workflow of choice:
1) deterministic but slow builds with Pipenv
2) or, minimalistic and fast installation with `pip`.
The only drawback is that `requirements.txt` needs to be changed every time `Pipfile.lock` changes, but `Pipenv` does all the hard work. :-)

This isn't intended as an attack on Kenneth Reitz's work. IMO, Pipenv is great for deterministic builds, but falls short on some use cases, and there are valid reasons why a user may deliberately decide not to use it in favor of a more traditional pipeline.

**Comment by pietrodn** [comment]
I would also include macOS installation instructions (see #142).

**Comment by bkueng** [comment]
Thanks for the extensive feedback @pietrodn! A combination with generating `requirements.txt` makes sense then. @mrpollo what do you think?

**Comment by mrpollo** [comment]
Thanks for the feedback @pietrodn, I'll add the `requirements.txt` and instructions on how to generate the file, we should try to add a test if we can figure out how, this will ensure we don't push a lock w/o an update to the txt file

**Comment by pietrodn** [comment]
We could make Pipenv generate a new `requirements.txt`, run `diff` against the one present in the working directory, and check the exit code.

**Comment by mrpollo** [comment]
I think I addressed the comments in my recent commits, please review

**Comment by pietrodn** [comment]
Looks good!

I tried the installation with both methods (venv + pip and pipenv), both with cached dependency downloads.

For pip I used this script (`install.sh`):
```bash
#/bin/bash

python3 -m venv venv
venv/bin/pip install -r requirements.txt
```

For pipenv, I used `pipenv sync --python=3.7`.

Here are the performances, measured with the `time` utility:

    ./install.sh              19,11s user 6,43s system 84% cpu 30,399 total
    pipenv sync --python=3.7  36,55s user 9,82s system 196% cpu 23,573 total

Pipenv runs faster by exploiting multiple CPUs.
`pipenv install` takes approximately the same time.

After removing `Pipfile.lock`, however, `pipenv install` takes forever (I launched it like 15 minutes ago and it's still running). Dependency locking is *very* time-consuming.

File: run_tests.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+set -e
+
+# Check requirements lock file
+pipenv lock --requirements > local-requirements.txt
+cmp --silent requirements.txt local-requirements.txt || echo "ERROR: Pipenv.lock difers from requirements.txt"
+
+# Check pylint
+# TODO: fix pylint on source and enable
+#bash run_pylint.sh
```
Comment (bkueng) [inline run_tests.sh]: Why is this disabled?

**Comment by mrpollo** [comment]
thanks for the metrics @pietrodn, I think we should always provide a lock if we are to move forward with this workflow change to `pipenv`.

My next goal is to make this a proper python package and expose flight_review as a command line tool to generate html reports out of ulog, I'll open an issue to discuss

**Pull request merged**